staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -107,14 +133,34 @@ func newWatcher(client *clientv3.Client, codec runtime.Codec, groupResource sche
 // If rev is zero, it will return the existing object(s) and then start watching from
 // the maximum revision+1 from returned objects.
 // If rev is non-zero, it will watch events happened after given revision.
-// If recursive is false, it watches on given key.
-// If recursive is true, it watches any children and directories under the key, excluding the root key itself.
-// pred must be non-nil. Only if pred matches the change, it will be returned.
-func (w *watcher) Watch(ctx context.Context, key string, rev int64, recursive, progressNotify bool, transformer value.Transformer, pred storage.SelectionPredicate) (watch.Interface, error) {
-	if recursive && !strings.HasSuffix(key, "/") {
+// If opts.Recursive is false, it watches on given key.
+// If opts.Recursive is true, it watches any children and directories under the key, excluding the root key itself.
+// pred must be non-nil. Only if opts.Predicate matches the change, it will be returned.
+func (w *watcher) Watch(ctx context.Context, key string, rev int64, transformer value.Transformer, opts storage.ListOptions) (watch.Interface, error) {
+	if opts.Recursive && !strings.HasSuffix(key, "/") {
 		key += "/"
 	}
-	wc := w.createWatchChan(ctx, key, rev, recursive, progressNotify, transformer, pred)
+	var initialEventsEndBookmarkRequired bool
+	progressNotify := opts.ProgressNotify
+	if utilfeature.DefaultFeatureGate.Enabled(features.WatchList) &&
+		opts.SendInitialEvents != nil &&
+		*opts.SendInitialEvents &&
+		opts.Predicate.AllowWatchBookmarks {
+		if w.newFunc == nil {
+			return nil, apierrors.NewInvalid(
+				schema.GroupKind{Group: w.groupResource.Group, Kind: w.groupResource.Resource},
+				"",
+				field.ErrorList{field.Forbidden(field.NewPath("sendInitialEvents"), "for watch is unsupported by the etcd storage because no newFunc was provided")},
+			)
+		}
+		// since there is no way to directly set
+		// opts.ProgressNotify from the API
+		// we need to take into account the value
+		// stored at opts.Predicate.AllowWatchBookmarks
+		progressNotify = true
```

Comment:
this is slightly confusing, we could rename it to `allowWatchBookmarks` at for the `watchChan`